### PR TITLE
`ultralytics 8.3.121` Add FocalLoss multi-class support

### PR DESCRIPTION
**Pull request by pow3rpi**
**Issue**: #20368

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved the Focal Loss function to support both single and per-class weighting for better handling of class imbalance. ⚖️✨

### 📊 Key Changes
- Updated the Focal Loss implementation to accept `alpha` as either a single value or a list, allowing per-class balancing.
- Converted `alpha` to a tensor for flexible computation and device compatibility.
- Minor docstring updates to clarify the new `alpha` parameter behavior.
- Bumped the Ultralytics package version to 8.3.121.

### 🎯 Purpose & Impact
- Enables more precise control over class imbalance by allowing users to specify different weights for each class.
- Improves model training, especially for datasets with uneven class distributions.
- Enhances flexibility and usability for advanced users working with custom datasets.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pow3rpi, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process. An Ultralytics engineer will also assist you soon.

To ensure a smooth integration of your contribution, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and follow the project's conventions.
- ✅ **Synchronize with Source**: Make sure your PR is up to date with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Please verify that all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, address the issues accordingly.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: If this is your first Ultralytics PR, ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For additional details, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If your PR is related to a bug and you haven't provided a minimum reproducible example (MRE), please add one to help us review and test your contribution efficiently.

Thank you for helping improve Ultralytics! 🚀

**Comment by pow3rpi** [comment]
@glenn-jocher @Y-T-G Tell me please if you want me to add any documentation regarding this PR. If yes, please show where to do it.

**Comment by Laughing-q** [comment]
@pow3rpi Hey thanks for the PR! However I think we might not want the changes there about `softmax`, I understand you might want to set different alpha value for different classes during loss calculation, can we have this additional feature while sticking with the `sigmoid` for the alpha calculation?
pic-250428-1415-12

**Comment by pow3rpi** [comment]
@Laughing-q Thank you for your review! Actually, softmax was suggested by @glenn-jocher in the issue discussion mentioned in the description. Could you please explain your point about softmax in detail?

**Comment by Laughing-q** [comment]
@pow3rpi I think we could stick with this part for `p_t` calculation: https://github.com/ultralytics/ultralytics/blob/20196b57d0285856d2b9e9ef095beb39d0643c8e/ultralytics/utils/loss.py#L67-L70, and remove the if-else branch for `softmax` as showed in above screenshot, just need to add handling when passed `alpha` argument is a list or `torch.Tensor`.

**Comment by pow3rpi** [comment]
@Laughing-q Could you please check the sigmoid version and tell me if you want me to change anything else?

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
@pow3rpi Hey I did some simplifications here, and tested it with following example and it works properly to me:
```python
import torch
import torch.nn as nn
import torch.optim as optim
from ultralytics.utils.loss import FocalLoss


# Dummy dataset for multi-class classification
class DummyDataset(torch.utils.data.Dataset):
    def __init__(self, num_samples=100, num_features=10, num_classes=3):
        self.num_samples = num_samples
        self.num_features = num_features
        self.num_classes = num_classes
        self.data = torch.randn(num_samples, num_features)  # Random features
        self.labels = torch.randint(0, num_classes, (num_samples,))  # Random multi-class labels

    def __len__(self):
        return self.num_samples

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]


# Simple model for multi-class classification
class SimpleModel(nn.Module):
    def __init__(self, input_size, num_classes):
        super(SimpleModel, self).__init__()
        self.fc = nn.Linear(input_size, num_classes)

    def forward(self, x):
        return self.fc(x)


# Initialize dataset, dataloader, model, loss, and optimizer
num_classes = 3
dataset = DummyDataset(num_classes=num_classes)
dataloader = torch.utils.data.DataLoader(dataset, batch_size=16, shuffle=True)

model = SimpleModel(input_size=10, num_classes=num_classes)
# criterion = FocalLoss(gamma=2.0, alpha=0.25)  # Using FocalLoss
criterion = FocalLoss(gamma=2.0, alpha=[0.25] * 3)  # Using FocalLoss
optimizer = optim.SGD(model.parameters(), lr=0.01)

# Training loop
for epoch in range(5):  # 5 epochs
    for data, labels in dataloader:
        optimizer.zero_grad()
        outputs = model(data)  # Raw logits
        labels_one_hot = nn.functional.one_hot(labels, num_classes=num_classes).float()  # Convert to one-hot
        loss = criterion(outputs, labels_one_hot)  # Compute focal loss
        loss.backward()
        optimizer.step()

    print(f"Epoch {epoch + 1}, Loss: {loss.item():.4f}")
```
Please let me know if it works to you as well! Thanks!

**Comment by Laughing-q** [comment]
@glenn-jocher This one is ready

**Comment by glenn-jocher** [comment]
Great, thank you @Laughing-q! @pow3rpi PR is approved and will likely merge tomorrow as part of our next release.

**Comment by UltralyticsAssistant** [comment]
🎉 Fantastic work, @pow3rpi, @Laughing-q, and @glenn-jocher! Your dedication to improving the Focal Loss function is a game-changer for users tackling class imbalance—empowering them with greater flexibility and precision.

As Helen Keller once said, "Alone we can do so little; together we can do so much." This collaboration truly embodies that spirit and pushes Ultralytics forward for everyone. Thank you for your impactful contributions and for making Ultralytics HUB and YOLO11 even better! 🚀

**Comment by charlesollion** [comment]
Hello @pow3rpi how would you use this one in practice?
I saw that the default Yolo11 segmentation task do not use `FocalLoss`, but `v8SegmentationLoss` which uses `BBoxLoss` (which includes `DFL` loss, somehow replacing a `FocalLoss`) + `bce`, i guess as an additional class loss (in `DFL`, classes are taken into account as well from my understanding).
I changed the `bce` by `FocalLoss` and kept the `DFL` as well to test your setup. I have two questions:
- is my approach sound?
- I use 0.25 for each normal class, how would you chose the values for unbalanced class, say classes with very little amount of labels?
Thanks!

**Comment by pow3rpi** [comment]
Hi @charlesollion! Thank you for your interest to this PR! I originally decided to add multi-class support to FocalLoss when I was using it in the detection task with v8DetectionLoss (YOLO 11 still uses this class for the loss) but, as far as I know, usage of it in the segmentation task is almost the same. 

1) Your approach sounds good bc in detection task, yes, you replace **self.bce** property with your FocalLoss and everything works as you want.

2) To calculate class weights basing on the number of detections for each class in the train set, I used the following code:
```
folder_path = f'{custom_base_path}/train/labels'
class_counts = defaultdict(int)
counts = []

for filename in os.listdir(folder_path):
  if filename.endswith('.txt'):
    file_path = os.path.join(folder_path, filename)
    with open(file_path, 'r') as f:
      for line in f:
        if line.strip():
          class_id = line.strip().split()[0]
          class_counts[class_id] += 1

print('Class occurrence counts:')
for class_id, count in sorted(class_counts.items(), key=lambda x: int(x[0])):
    print(f'Class {class_id}: {count}')
    counts.append(count)
print()
print(counts)

total = sum(counts)

alpha = [total / c for c in counts]
alpha = torch.tensor(alpha, dtype=torch.float32)
alpha = alpha / alpha.sum()  # normalize (optional)

print('Alpha vector:', alpha)
```
, so you can adjust it to your case. 


Finally, talking about dfl loss, first of all, you should make sure that it is actually used in the YOLO version you are using as the official description says:

> dfl - Weight of the distribution focal loss, used in certain YOLO versions for fine-grained classification.

Secondly, to understand dfl loss better you could refer to the issue #4219.

I hope my answer will help you. If you have any other questions don't hesitate to reach me out again.
Good luck!

**Comment by charlesollion** [comment]
Thanks for your reply!
I might be mistaken, but i don't see anywhere in yolov8 or 11 where the `FocalLoss` would be used, in current main branch.
In tasks.py, the loss is defined for the model as either `E2EDetectLoss` or `v8DetectionLoss`, and the first one uses `v8DetectionLoss`.
In `v8DetectionLoss`, there's bce and dfl loss, I don't see the usage of `FocalLoss`, except in `DETRLoss`.

**Comment by pow3rpi** [comment]
@charlesollion In th ecurrent version you need to use FocalLoss yourself where needed by editing source code. Now, I am working on making loss function configurable.
